File: frontend/src/api/datatypes.ts
// ============================================================
// 标准化数据类型定义
// ============================================================

export type StandardDataType =
  | 'number'
  | 'integer'
  | 'boolean'
  | 'string'
  | 'date'
  | 'datetime'
  | 'array'
  | 'map'
  | 'json'
  | 'unknown';

export interface TypeConfig {
  precision?: number;
  scale?: number;
  itemType?: StandardDataType;
  keyType?: StandardDataType;
  valueType?: StandardDataType;
}

export interface DatasetColumn {
  name: string;
  expr?: string;
  type: StandardDataType;
  typeConfig?: TypeConfig;
  comment?: string;
  role: 'dimension' | 'metric';
  isVirtual?: boolean;
}

export interface VirtualField {
  name: string;
  expression: string;
  exprType: ExpressionType;
  resultType: StandardDataType;
  config?: TypeConfig;
}

export type ExpressionType =
  | 'arithmetic'
  | 'string'
  | 'datetime'
  | 'conditional'
  | 'aggregate'
  | 'custom';

// ============================================================
// 类型映射配置
// ============================================================

export interface TypeMappingRule {
  standard: StandardDataType;
  config?: TypeConfig;
}

export type TypeMapping = Record<string, TypeMappingRule>;

export const starRocksMapping: TypeMapping = {
  tinyint: { standard: 'integer' },
  smallint: { standard: 'integer' },
  int: { standard: 'integer' },
  bigint: { standard: 'integer' },
  largeint: { standard: 'integer' },
  float: { standard: 'number' },
  double: { standard: 'number' },
  decimal: { standard: 'number' },
  bool: { standard: 'boolean' },
  boolean: { standard: 'boolean' },
  varchar: { standard: 'string' },
  string: { standard: 'string' },
  char: { standard: 'string' },
  date: { standard: 'date' },
  datetime: { standard: 'datetime' },
  timestamp: { standard: 'datetime' },
  array: { standard: 'array' },
  map: { standard: 'map' },
  json: { standard: 'json' },
};

export const postgresqlMapping: TypeMapping = {
  smallint: { standard: 'integer' },
  integer: { standard: 'integer' },
  bigint: { standard: 'integer' },
  real: { standard: 'number' },
  double: { standard: 'number' },
  'double precision': { standard: 'number' },
  numeric: { standard: 'number' },
  decimal: { standard: 'number' },
  boolean: { standard: 'boolean' },
  bool: { standard: 'boolean' },
  varchar: { standard: 'string' },
  text: { standard: 'string' },
  char: { standard: 'string' },
  date: { standard: 'date' },
  timestamp: { standard: 'datetime' },
  timestamptz: { standard: 'datetime' },
  json: { standard: 'json' },
  jsonb: { standard: 'json' },
};

export const mysqlMapping: TypeMapping = {
  tinyint: { standard: 'integer' },
  smallint: { standard: 'integer' },
  mediumint: { standard: 'integer' },
  int: { standard: 'integer' },
  integer: { standard: 'integer' },
  bigint: { standard: 'integer' },
  float: { standard: 'number' },
  double: { standard: 'number' },
  real: { standard: 'number' },
  decimal: { standard: 'number' },
  dec: { standard: 'number' },
  numeric: { standard: 'number' },
  fixed: { standard: 'number' },
  bool: { standard: 'boolean' },
  boolean: { standard: 'boolean' },
  'tinyint(1)': { standard: 'boolean' },
  char: { standard: 'string' },
  varchar: { standard: 'string' },
  tinytext: { standard: 'string' },
  text: { standard: 'string' },
  mediumtext: { standard: 'string' },
  longtext: { standard: 'string' },
  date: { standard: 'date' },
  datetime: { standard: 'datetime' },
  timestamp: { standard: 'datetime' },
  json: { standard: 'json' },
};

export const clickhouseMapping: TypeMapping = {
  int8: { standard: 'integer' },
  int16: { standard: 'integer' },
  int32: { standard: 'integer' },
  int64: { standard: 'integer' },
  int128: { standard: 'integer' },
  int256: { standard: 'integer' },
  uint8: { standard: 'integer' },
  uint16: { standard: 'integer' },
  uint32: { standard: 'integer' },
  uint64: { standard: 'integer' },
  uint128: { standard: 'integer' },
  uint256: { standard: 'integer' },
  float32: { standard: 'number' },
  float64: { standard: 'number' },
  decimal: { standard: 'number' },
  bool: { standard: 'boolean' },
  string: { standard: 'string' },
  fixedstring: { standard: 'string' },
  uuid: { standard: 'string' },
  date: { standard: 'date' },
  date32: { standard: 'date' },
  datetime: { standard: 'datetime' },
  datetime64: { standard: 'datetime' },
  array: { standard: 'array' },
  map: { standard: 'map' },
  json: { standard: 'json' },
  object: { standard: 'json' },
};

export const datasourceTypeMappings: Record<string, TypeMapping> = {
  starrocks: starRocksMapping,
  postgresql: postgresqlMapping,
  mysql: mysqlMapping,
  clickhouse: clickhouseMapping,
};

// ============================================================
// 类型转换函数
// ============================================================

export function toStandardType(
  sourceType: string,
  datasourceType: string
): StandardDataType {
  const mapping = datasourceTypeMappings[datasourceType.toLowerCase()];
  if (!mapping) {
    return 'unknown';
  }

  const normalizedType = sourceType.toLowerCase().trim();
  
  // 特殊处理带参数的复杂类型
  if (normalizedType.startsWith('decimal')) return 'number';
  if (normalizedType.startsWith('numeric')) return 'number';
  if (normalizedType.startsWith('array')) return 'array';
  if (normalizedType.startsWith('map')) return 'map';
  if (normalizedType.startsWith('varchar')) return 'string';
  if (normalizedType.startsWith('char')) return 'string';
  if (normalizedType.startsWith('enum')) return 'string';
  if (normalizedType.startsWith('set')) return 'string';
  if (normalizedType.startsWith('fixedstring')) return 'string';

  const rule = mapping[normalizedType];
  return rule?.standard ?? 'unknown';
}

export function toSourceType(
  stdType: StandardDataType,
  datasourceType: string,
  config?: TypeConfig
): string {
  const mapping: Record<string, Record<string, string>> = {
    starrocks: {
      number: config?.precision ? `decimal(${config.precision},${config.scale ?? 0})` : 'double',
      integer: 'bigint',
      boolean: 'boolean',
      string: 'varchar',
      date: 'date',
      datetime: 'datetime',
      array: 'array',
      map: 'map',
      json: 'json',
    },
    postgresql: {
      number: config?.precision ? `numeric(${config.precision},${config.scale ?? 0})` : 'numeric',
      integer: 'bigint',
      boolean: 'boolean',
      string: 'varchar',
      date: 'date',
      datetime: 'timestamp',
      array: 'array',
      map: 'jsonb',
      json: 'jsonb',
    },
    mysql: {
      number: config?.precision ? `decimal(${config.precision},${config.scale ?? 0})` : 'decimal',
      integer: 'bigint',
      boolean: 'tinyint(1)',
      string: 'varchar',
      date: 'date',
      datetime: 'datetime',
      array: 'json',
      map: 'json',
      json: 'json',
    },
    clickhouse: {
      number: config?.precision ? `decimal(${config.precision},${config.scale ?? 0})` : 'float64',
      integer: 'int64',
      boolean: 'bool',
      string: 'string',
      date: 'date',
      datetime: 'datetime',
      array: 'array',
      map: 'map',
      json: 'json',
    },
  };

  const sourceMapping = mapping[datasourceType.toLowerCase()];
  if (!sourceMapping) {
    return 'varchar';
  }

  return sourceMapping[stdType] ?? 'varchar';
}

// ============================================================
// 类型辅助函数
// ============================================================

export const dataTypeOptions: Array<{ value: StandardDataType; label: string; category: string }> = [
  { value: 'number', label: '数值', category: '数值' },
  { value: 'integer', label: '整数', category: '数值' },
  { value: 'boolean', label: '布尔值', category: '数值' },
  { value: 'string', label: '字符串', category: '文本' },
  { value: 'date', label: '日期', category: '日期时间' },
  { value: 'datetime', label: '日期时间', category: '日期时间' },
  { value: 'array', label: '数组', category: '复杂类型' },
  { value: 'map', label: '字典', category: '复杂类型' },
  { value: 'json', label: 'JSON', category: '复杂类型' },
];

export function getTypeCategory(stdType: StandardDataType): string {
  const categories: Record<string, string> = {
    number: '数值',
    integer: '数值',
    boolean: '数值',
    string: '文本',
    date: '日期时间',
    datetime: '日期时间',
    array: '复杂类型',
    map: '复杂类型',
    json: '复杂类型',
  };
  return categories[stdType] ?? '未知';
}

export function isNumericType(stdType: StandardDataType): boolean {
  return stdType === 'number' || stdType === 'integer';
}

export function isDateTimeType(stdType: StandardDataType): boolean {
  return stdType === 'date' || stdType === 'datetime';
}

export function isComplexType(stdType: StandardDataType): boolean {
  return stdType === 'array' || stdType === 'map' || stdType === 'json';
}

// ============================================================
// 表达式推断
// ============================================================

const expressionFunctions: Record<string, StandardDataType> = {
  SUM: 'number',
  AVG: 'number',
  COUNT: 'integer',
  MAX: 'number',
  MIN: 'number',
  CONCAT: 'string',
  UPPER: 'string',
  LOWER: 'string',
  TRIM: 'string',
  SUBSTRING: 'string',
  LENGTH: 'integer',
  YEAR: 'integer',
  MONTH: 'integer',
  DAY: 'integer',
  DATE_ADD: 'datetime',
  DATE_SUB: 'datetime',
  DATEDIFF: 'integer',
  ROUND: 'number',
  ABS: 'number',
  FLOOR: 'integer',
  CEIL: 'integer',
  POWER: 'number',
  MOD: 'integer',
};

export function inferExpressionType(expression: string): StandardDataType {
  const upperExpr = expression.toUpperCase();

  for (const [func, type] of Object.entries(expressionFunctions)) {
    if (upperExpr.includes(func)) {
      return type;
    }
  }

  // 算术运算
  if (/[+\-*/%]/.test(upperExpr)) {
    if (upperExpr.includes('||')) return 'string';
    return 'number';
  }

  return 'unknown';
}

File: backend/internal/model/datatypes.go
package model

import (
	"fmt"
	"strings"
)

// ============================================================
// 标准化数据类型定义
// ============================================================

// StandardDataType 定义标准化数据类型枚举
type StandardDataType string

const (
	// 数值类型
	TypeNumber  StandardDataType = "number"  // 数值（含小数）
	TypeInteger StandardDataType = "integer" // 整数
	TypeBoolean StandardDataType = "boolean" // 布尔值

	// 文本类型
	TypeString StandardDataType = "string" // 字符串

	// 日期时间类型
	TypeDate     StandardDataType = "date"     // 日期
	TypeDateTime StandardDataType = "datetime" // 日期时间

	// 复杂类型
	TypeArray StandardDataType = "array" // 数组
	TypeMap   StandardDataType = "map"   // 字典
	TypeJSON  StandardDataType = "json"  // JSON

	// 未知类型
	TypeUnknown StandardDataType = "unknown"
)

// StandardDataTypes 所有支持的标准化类型
var StandardDataTypes = []StandardDataType{
	TypeNumber,
	TypeInteger,
	TypeBoolean,
	TypeString,
	TypeDate,
	TypeDateTime,
	TypeArray,
	TypeMap,
	TypeJSON,
}

// TypeCategory 类型的分类
var TypeCategory = map[StandardDataType]string{
	TypeNumber:   "数值",
	TypeInteger:  "数值",
	TypeBoolean:  "数值",
	TypeString:   "文本",
	TypeDate:     "日期时间",
	TypeDateTime: "日期时间",
	TypeArray:    "复杂类型",
	TypeMap:      "复杂类型",
	TypeJSON:     "复杂类型",
}

// ============================================================
// 类型配置
// ============================================================

// TypeConfig 类型配置（用于精度控制）
type TypeConfig struct {
	Precision int              `json:"precision,omitempty"`  // 总位数（如 10）
	Scale     int              `json:"scale,omitempty"`      // 小数位数（如 2）
	ItemType  StandardDataType `json:"item_type,omitempty"`  // 数组元素类型
	KeyType   StandardDataType `json:"key_type,omitempty"`   // 字典键类型
	ValueType StandardDataType `json:"value_type,omitempty"` // 字典值类型
}

// GetDisplayName 获取类型的显示名称
func (t StandardDataType) GetDisplayName() string {
	names := map[StandardDataType]string{
		TypeNumber:   "数值",
		TypeInteger:  "整数",
		TypeBoolean:  "布尔值",
		TypeString:   "字符串",
		TypeDate:     "日期",
		TypeDateTime: "日期时间",
		TypeArray:    "数组",
		TypeMap:      "字典",
		TypeJSON:     "JSON",
	}
	if name, ok := names[t]; ok {
		return name
	}
	return string(t)
}

// IsNumeric 判断是否为数值类型
func (t StandardDataType) IsNumeric() bool {
	return t == TypeNumber || t == TypeInteger
}

// IsDateTime 判断是否为日期时间类型
func (t StandardDataType) IsDateTime() bool {
	return t == TypeDate || t == TypeDateTime
}

// IsComplex 判断是否为复杂类型
func (t StandardDataType) IsComplex() bool {
	return t == TypeArray || t == TypeMap || t == TypeJSON
}

// ============================================================
// 数据源类型映射器接口
// ============================================================

// DataTypeMapper 数据类型映射器接口
type DataTypeMapper interface {
	// ToStandard 将数据源类型转换为标准类型
	ToStandard(sourceType string) (StandardDataType, TypeConfig, error)

	// ToSource 将标准类型转换为数据源类型
	ToSource(stdType StandardDataType, config TypeConfig) string

	// GetSourceName 获取数据源名称
	GetSourceName() string
}

// ============================================================
// StarRocks 类型映射器
// ============================================================

// StarRocksMapper StarRocks 类型映射器
type StarRocksMapper struct{}

var starRocksTypeMap = map[string]StandardDataType{
	// 数值类型
	"tinyint":   TypeInteger,
	"smallint":  TypeInteger,
	"int":       TypeInteger,
	"bigint":    TypeInteger,
	"largeint":  TypeInteger,
	"float":     TypeNumber,
	"double":    TypeNumber,
	"decimal":   TypeNumber,
	"decimalv2": TypeNumber,

	// 布尔类型
	"bool":    TypeBoolean,
	"boolean": TypeBoolean,

	// 字符串类型
	"char":      TypeString,
	"varchar":   TypeString,
	"string":    TypeString,
	"binary":    TypeString,
	"varbinary": TypeString,

	// 日期时间类型
	"date":      TypeDate,
	"datetime":  TypeDateTime,
	"timestamp": TypeDateTime,

	// 复杂类型
	"array": TypeArray,
	"map":   TypeMap,
	"json":  TypeJSON,
}

func (m *StarRocksMapper) ToStandard(sourceType string) (StandardDataType, TypeConfig, error) {
	sourceType = strings.ToLower(strings.TrimSpace(sourceType))

	// 处理带参数的数值类型 (如 int(11), tinyint(2))
	// 先提取基础类型
	if idx := strings.Index(sourceType, "("); idx > 0 {
		baseType := sourceType[:idx]
		// 检查基础类型是否在映射表中
		if stdType, ok := starRocksTypeMap[baseType]; ok {
			return stdType, TypeConfig{}, nil
		}
	}

	// 处理带参数的复杂类型
	if strings.HasPrefix(sourceType, "array<") {
		return TypeArray, TypeConfig{}, nil
	}
	if strings.HasPrefix(sourceType, "map<") {
		return TypeMap, TypeConfig{}, nil
	}

	// 处理 decimal(p,s)
	if strings.HasPrefix(sourceType, "decimal") {
		return TypeNumber, TypeConfig{}, nil
	}

	stdType, ok := starRocksTypeMap[sourceType]
	if !ok {
		return TypeUnknown, TypeConfig{}, fmt.Errorf("unknown StarRocks type: %s", sourceType)
	}
	return stdType, TypeConfig{}, nil
}

func (m *StarRocksMapper) ToSource(stdType StandardDataType, config TypeConfig) string {
	switch stdType {
	case TypeNumber:
		if config.Precision > 0 && config.Scale > 0 {
			return fmt.Sprintf("decimal(%d,%d)", config.Precision, config.Scale)
		}
		return "double"
	case TypeInteger:
		return "bigint"
	case TypeBoolean:
		return "boolean"
	case TypeString:
		return "varchar"
	case TypeDate:
		return "date"
	case TypeDateTime:
		return "datetime"
	case TypeArray:
		return "array"
	case TypeMap:
		return "map"
	case TypeJSON:
		return "json"
	default:
		return "varchar"
	}
}

func (m *StarRocksMapper) GetSourceName() string {
	return "starrocks"
}

// ============================================================
// PostgreSQL 类型映射器
// ============================================================

// PostgreSQLMapper PostgreSQL 类型映射器
type PostgreSQLMapper struct{}

var postgresTypeMap = map[string]StandardDataType{
	// 数值类型
	"smallint":         TypeInteger,
	"smallserial":      TypeInteger,
	"integer":          TypeInteger,
	"serial":           TypeInteger,
	"bigint":           TypeInteger,
	"bigserial":        TypeInteger,
	"real":             TypeNumber,
	"double":           TypeNumber,
	"double precision": TypeNumber,
	"numeric":          TypeNumber,
	"decimal":          TypeNumber,

	// 布尔类型
	"boolean": TypeBoolean,
	"bool":    TypeBoolean,

	// 字符串类型
	"char":              TypeString,
	"character":         TypeString,
	"varchar":           TypeString,
	"character varying": TypeString,
	"text":              TypeString,

	// 日期时间类型
	"date":        TypeDate,
	"time":        TypeString,
	"timestamp":   TypeDateTime,
	"timestamptz": TypeDateTime,
	"timetz":      TypeDateTime,

	// 复杂类型
	"array": TypeArray,
	"json":  TypeJSON,
	"jsonb": TypeJSON,
}

func (m *PostgreSQLMapper) ToStandard(sourceType string) (StandardDataType, TypeConfig, error) {
	sourceType = strings.ToLower(strings.TrimSpace(sourceType))

	// 处理 array 类型
	if strings.HasPrefix(sourceType, "array") {
		return TypeArray, TypeConfig{}, nil
	}

	// 处理 numeric(p,s)
	if strings.HasPrefix(sourceType, "numeric") || strings.HasPrefix(sourceType, "decimal") {
		return TypeNumber, TypeConfig{}, nil
	}

	stdType, ok := postgresTypeMap[sourceType]
	if !ok {
		return TypeUnknown, TypeConfig{}, fmt.Errorf("unknown PostgreSQL type: %s", sourceType)
	}
	return stdType, TypeConfig{}, nil
}

func (m *PostgreSQLMapper) ToSource(stdType StandardDataType, config TypeConfig) string {
	switch stdType {
	case TypeNumber:
		if config.Precision > 0 {
			return fmt.Sprintf("numeric(%d,%d)", config.Precision, config.Scale)
		}
		return "numeric"
	case TypeInteger:
		return "bigint"
	case TypeBoolean:
		return "boolean"
	case TypeString:
		return "varchar"
	case TypeDate:
		return "date"
	case TypeDateTime:
		return "timestamp"
	case TypeArray:
		return "array"
	case TypeMap:
		return "jsonb"
	case TypeJSON:
		return "jsonb"
	default:
		return "varchar"
	}
}

func (m *PostgreSQLMapper) GetSourceName() string {
	return "postgresql"
}

// ============================================================
// MySQL 类型映射器
// ============================================================

// MySQLMapper MySQL 类型映射器
type MySQLMapper struct{}

var mysqlTypeMap = map[string]StandardDataType{
	// 数值类型
	"tinyint":   TypeInteger,
	"smallint":  TypeInteger,
	"mediumint": TypeInteger,
	"int":       TypeInteger,
	"integer":   TypeInteger,
	"bigint":    TypeInteger,
	"float":     TypeNumber,
	"double":    TypeNumber,
	"real":      TypeNumber,
	"decimal":   TypeNumber,
	"dec":       TypeNumber,
	"numeric":   TypeNumber,
	"fixed":     TypeNumber,

	// 布尔类型（MySQL 用 tinyint(1) 表示）
	"bool":       TypeBoolean,
	"boolean":    TypeBoolean,
	"tinyint(1)": TypeBoolean,

	// 字符串类型
	"char":       TypeString,
	"varchar":    TypeString,
	"tinytext":   TypeString,
	"text":       TypeString,
	"mediumtext": TypeString,
	"longtext":   TypeString,
	"binary":     TypeString,
	"varbinary":  TypeString,
	"blob":       TypeString,
	"enum":       TypeString,
	"set":        TypeString,

	// 日期时间类型
	"date":      TypeDate,
	"datetime":  TypeDateTime,
	"timestamp": TypeDateTime,
	"time":      TypeString,
	"year":      TypeInteger,

	// 复杂类型
	"json":  TypeJSON,
	"array": TypeArray, // MySQL 8.0+ 支持
}

func (m *MySQLMapper) ToStandard(sourceType string) (StandardDataType, TypeConfig, error) {
	sourceType = strings.ToLower(strings.TrimSpace(sourceType))

	// 处理 tinyint(1) 作为布尔
	if sourceType == "tinyint(1)" {
		return TypeBoolean, TypeConfig{}, nil
	}

	// 处理 decimal(p,s) 或 numeric(p,s)
	if strings.HasPrefix(sourceType, "decimal") || strings.HasPrefix(sourceType, "numeric") || strings.HasPrefix(sourceType, "dec") || strings.HasPrefix(sourceType, "fixed") {
		return TypeNumber, TypeConfig{}, nil
	}

	// 处理 enum 和 set
	if strings.HasPrefix(sourceType, "enum") || strings.HasPrefix(sourceType, "set") {
		return TypeString, TypeConfig{}, nil
	}

	// 处理 char(n)
	if strings.HasPrefix(sourceType, "char") {
		return TypeString, TypeConfig{}, nil
	}

	// 处理 varchar(n)
	if strings.HasPrefix(sourceType, "varchar") {
		return TypeString, TypeConfig{}, nil
	}

	stdType, ok := mysqlTypeMap[sourceType]
	if !ok {
		return TypeUnknown, TypeConfig{}, fmt.Errorf("unknown MySQL type: %s", sourceType)
	}
	return stdType, TypeConfig{}, nil
}

func (m *MySQLMapper) ToSource(stdType StandardDataType, config TypeConfig) string {
	switch stdType {
	case TypeNumber:
		if config.Precision > 0 && config.Scale > 0 {
			return fmt.Sprintf("decimal(%d,%d)", config.Precision, config.Scale)
		}
		return "decimal"
	case TypeInteger:
		return "bigint"
	case TypeBoolean:
		return "tinyint(1)"
	case TypeString:
		return "varchar"
	case TypeDate:
		return "date"
	case TypeDateTime:
		return "datetime"
	case TypeArray:
		return "json"
	case TypeMap:
		return "json"
	case TypeJSON:
		return "json"
	default:
		return "varchar"
	}
}

func (m *MySQLMapper) GetSourceName() string {
	return "mysql"
}

// ============================================================
// ClickHouse 类型映射器
// ============================================================

// ClickHouseMapper ClickHouse 类型映射器
type ClickHouseMapper struct{}

var clickHouseTypeMap = map[string]StandardDataType{
	// 数值类型
	"int8":    TypeInteger,
	"int16":   TypeInteger,
	"int32":   TypeInteger,
	"int64":   TypeInteger,
	"int128":  TypeInteger,
	"int256":  TypeInteger,
	"uint8":   TypeInteger,
	"uint16":  TypeInteger,
	"uint32":  TypeInteger,
	"uint64":  TypeInteger,
	"uint128": TypeInteger,
	"uint256": TypeInteger,
	"float32": TypeNumber,
	"float64": TypeNumber,
	"decimal": TypeNumber,

	// 布尔类型
	"bool": TypeBoolean,

	// 字符串类型
	"string":      TypeString,
	"fixedstring": TypeString,
	"uuid":        TypeString,
	"ipv4":        TypeString,
	"ipv6":        TypeString,

	// 日期时间类型
	"date":       TypeDate,
	"date32":     TypeDate,
	"datetime":   TypeDateTime,
	"datetime64": TypeDateTime,

	// 复杂类型
	"array":          TypeArray,
	"map":            TypeMap,
	"tuple":          TypeArray,
	"nested":         TypeMap,
	"json":           TypeJSON,
	"object":         TypeJSON,
	"object('json')": TypeJSON,
}

func (m *ClickHouseMapper) ToStandard(sourceType string) (StandardDataType, TypeConfig, error) {
	sourceType = strings.ToLower(strings.TrimSpace(sourceType))

	// 处理 decimal(p,s)
	if strings.HasPrefix(sourceType, "decimal") {
		return TypeNumber, TypeConfig{}, nil
	}

	// 处理 array(T)
	if strings.HasPrefix(sourceType, "array") {
		return TypeArray, TypeConfig{}, nil
	}

	// 处理 map(K,V)
	if strings.HasPrefix(sourceType, "map") {
		return TypeMap, TypeConfig{}, nil
	}

	// 处理 FixedString(n)
	if strings.HasPrefix(sourceType, "fixedstring") {
		return TypeString, TypeConfig{}, nil
	}

	// 处理 Tuple
	if strings.HasPrefix(sourceType, "tuple") {
		return TypeArray, TypeConfig{}, nil
	}

	// 处理 DateTime64
	if strings.HasPrefix(sourceType, "datetime64") {
		return TypeDateTime, TypeConfig{}, nil
	}

	stdType, ok := clickHouseTypeMap[sourceType]
	if !ok {
		return TypeUnknown, TypeConfig{}, fmt.Errorf("unknown ClickHouse type: %s", sourceType)
	}
	return stdType, TypeConfig{}, nil
}

func (m *ClickHouseMapper) ToSource(stdType StandardDataType, config TypeConfig) string {
	switch stdType {
	case TypeNumber:
		if config.Precision > 0 && config.Scale > 0 {
			return fmt.Sprintf("decimal(%d,%d)", config.Precision, config.Scale)
		}
		return "float64"
	case TypeInteger:
		return "int64"
	case TypeBoolean:
		return "bool"
	case TypeString:
		return "string"
	case TypeDate:
		return "date"
	case TypeDateTime:
		return "datetime"
	case TypeArray:
		return "array"
	case TypeMap:
		return "map"
	case TypeJSON:
		return "json"
	default:
		return "string"
	}
}

func (m *ClickHouseMapper) GetSourceName() string {
	return "clickhouse"
}

// ============================================================
// 工厂函数
// ============================================================

// NewDataTypeMapper 创建数据类型映射器
func NewDataTypeMapper(sourceType string) (DataTypeMapper, error) {
	switch strings.ToLower(sourceType) {
	case "starrocks":
		return &StarRocksMapper{}, nil
	case "postgresql", "postgres":
		return &PostgreSQLMapper{}, nil
	case "mysql":
		return &MySQLMapper{}, nil
	case "clickhouse":
		return &ClickHouseMapper{}, nil
	default:
		return nil, fmt.Errorf("unsupported datasource type: %s", sourceType)
	}
}

// ============================================================
// 虚拟字段表达式
// ============================================================

// ExpressionType 表达式类型
type ExpressionType string

const (
	ExprArithmetic  ExpressionType = "arithmetic"  // 算术运算
	ExprString      ExpressionType = "string"      // 字符串操作
	ExprDateTime    ExpressionType = "datetime"    // 日期时间操作
	ExprConditional ExpressionType = "conditional" // 条件判断
	ExprAggregate   ExpressionType = "aggregate"   // 聚合函数
	ExprCustom      ExpressionType = "custom"      // 自定义 SQL
)

// VirtualField 虚拟字段定义
type VirtualField struct {
	Name       string           `json:"name"`        // 字段名称
	Expression string           `json:"expression"`  // 表达式（如：price * quantity）
	ExprType   ExpressionType   `json:"expr_type"`   // 表达式类型
	ResultType StandardDataType `json:"result_type"` // 结果数据类型
	Config     TypeConfig       `json:"config"`      // 类型配置
}

// SupportedExpression 支持的表达式函数
var SupportedExpressions = map[string]StandardDataType{
	// 聚合函数
	"SUM":   TypeNumber,
	"AVG":   TypeNumber,
	"COUNT": TypeInteger,
	"MAX":   TypeNumber,
	"MIN":   TypeNumber,

	// 字符串操作
	"CONCAT":    TypeString,
	"UPPER":     TypeString,
	"LOWER":     TypeString,
	"TRIM":      TypeString,
	"SUBSTRING": TypeString,
	"LENGTH":    TypeInteger,
	"LENGTHB":   TypeInteger,

	// 日期时间操作
	"YEAR":          TypeInteger,
	"MONTH":         TypeInteger,
	"DAY":           TypeInteger,
	"DATE_ADD":      TypeDateTime,
	"DATE_SUB":      TypeDateTime,
	"DATEDIFF":      TypeInteger,
	"TIMESTAMPDIFF": TypeInteger,

	// 数值操作
	"ROUND": TypeNumber,
	"ABS":   TypeNumber,
	"FLOOR": TypeInteger,
	"CEIL":  TypeInteger,
	"POWER": TypeNumber,
	"MOD":   TypeInteger,

	// 条件判断
	"IF":       TypeUnknown, // 取决于返回值
	"CASE":     TypeUnknown,
	"COALESCE": TypeUnknown,
	"NULLIF":   TypeUnknown,
}

// InferExpressionResultType 推断表达式的结果类型
func InferExpressionResultType(expr string) StandardDataType {
	upperExpr := strings.ToUpper(expr)

	// 检查已知函数
	for funcName, resultType := range SupportedExpressions {
		if strings.Contains(upperExpr, funcName) {
			if resultType != TypeUnknown {
				return resultType
			}
		}
	}

	// 默认推断
	// 如果包含字符串连接操作符，推断为字符串
	if strings.Contains(expr, "||") {
		return TypeString
	}
	// 如果包含算术运算符，推断为数值
	arithmeticOps := []string{"+", "-", "*", "/", "%"}
	for _, op := range arithmeticOps {
		if strings.Contains(expr, op) {
			return TypeNumber
		}
	}

	// 如果是简单字段引用，无法推断
	return TypeUnknown
}
